staging/src/k8s.io/component-base/logs/json/json.go

```diff
@@ -50,18 +50,14 @@ func NewJSONLogger(infoStream, errorStream zapcore.WriteSyncer) (logr.Logger, fu
 		// The later is important when both streams get merged into a single
 		// stream by the consumer (same console for a command line tool, pod
 		// log for a container) because without it, messages get reordered.
-		flushError := writeWithFlushing{
-			WriteSyncer: errorStream,
-			other:       infoStream,
-		}
```

Comment:
Removing this is a loss of functionality that isn't described explicitly. Are you sure that this is better? If you remove the code, the comment also would need to be updated.

Reply:
I'm looking into implementing buffer flushing instead of syncing whole pipeline, however I don't think this would achieve proper ordering as described in comment. I don't think that keeping logs ordered was ever goal, it's good property for human readable logs. It's good that we maintain this property by default, however when splitting stream is enabled I don't see a reason to sacrifice throughput as most log processing solutions don't care about order. It's even expected that logs are processed out of order, for example I have seen this behavior in Fluentbit.

Reply:
I was thinking of the case where stdout and stderr get redirected into the same file or pipe and then get processed. But you are probably right, what matters in the end is just the time stamp attached to each message, not the order. So dropping the syncing between the two streams can indeed be dropped if we can't make it faster.

That it is this slow is a bit surprising. I'd like to understand better where the slowdown really comes from (mutex contention? slow Sync()).

Reply:
Another thought: when both streams go to the same file descriptor, can we guarantee that log messages are written atomically?

Otherwise we could end up with interleaved log messages (`<start of message 1><message 2><end of message 1>`).

Reply:
`writeWithFlushing` didn't handle that because no lock is held while writing.

Reply:
> Another thought: when both streams go to the same file descriptor, can we guarantee that log messages are written atomically?

I don't think so. It's up to log consumer to merge streams, if it not able to keep up with producer generating to much data to one of the streams it is expected to interleave streams to avoid starving the other one. This is what I expect to happen from bash pipe merging, there will be no interleaving as long as volume of logs is reasonable, however it cannot guarantee this in high load scenarios. 

The only way to handle it would be to have application log aware merger that will be able to recognize when single log message starts and end. We technically could implement such functionality in kube-log-runner, however I don't think it's worth it. As splitting streams is an optional feature, it's much more reasonable to document this behavior so users are aware of this behavior when enabling this feature.

Reply:
So we document that split stream assumes independent streams. Works for me.

I'm wondering whether we then want another mode: single stream with buffering of info messages. That might provide some performance benefit (to be measured) at the risk of loosing in-memory messages in case of a crash.